contents/functionalProgramming/intro.md

```diff
@@ -0,0 +1,132 @@
+
+```Python
+# Iterative
+def getSumOfListIter(listOfNumbers):
+    sum = 0
+    for number in listOfNumbers:
+        sum += number
+    return sum
+
+# Recursive
+def getSumOfListRec(listOfNumbers):
+    if len(listOfNumbers) == 0:
+        return 0
+    return listOfNumbers[0] + getSumOfListRec(listOfNumbers[1:])
+```
+
+Hence, to be able to write functional languages effectively, it means being able to replace Iteration with Recursion. Here are some guides to help you on that:
+
+- [Introduction to Recursion](https://www.topcoder.com/community/data-science/data-science-tutorials/an-introduction-to-recursion-part-1/)
+- [Recursion in functional programming](https://dzone.com/articles/functional-programming-recursion)
+
+#### Higher Order Functions
+
+Higher Order Functions are functions that either take functions as parameters, return a function or both. As functions are first class citizens in functional programming languages, this allows [functions to be passed around like objects and values](https://en.wikipedia.org/wiki/Functional_programming#First-class_and_higher-order_functions).
+
+One example of Higher Order Functions is `map`, which takes in a function and a collection of objects.
+
+```python
+length_of_names = map(len, ['Alex', 'Thomas', 'John'])
+print(length_of_names) # [4,6,4]
+```
+
+`map` takes in 2 parameters, a function and a collection. `map` returns a new collection whose elements are the result of applying the given function on the elements of the given collection. In the given example, `map` returns `[len('Alex'), len('Thomas'), len('John')]`
+
+There are useful functions and techniques that are based off Higher Order Functions in Functional programming.
+
+- [Currying](https://hackernoon.com/javascript-and-functional-programming-currying-pt-4-96e3230782ab)
+- [Map, Reduce and Filter](https://medium.freecodecamp.org/higher-order-functions-in-javascript-d9101f9cf528)
+
+## Advantages of Functional Programming
+
+The common question when learning about functional programming is what's the point of having us going through all the trouble of having pure functions, state immutability and absence of loops. Turns out, there are various positive implications of functional programming:
+
+- [More Bug-Free Code](https://www.quora.com/Are-software-written-using-Functional-Programming-less-buggy-more-robust-and-stable) for example famously in Haskell, [if it compiles it usually works](https://wiki.haskell.org/Why_Haskell_just_works).
+- [Concurrency](https://softwareengineering.stackexchange.com/questions/293851/what-is-it-about-functional-programming-that-makes-it-inherently-adapted-to-para). Having state immutability and absence of side effects can allow for bug-free concurrent code with less concerns of race conditions on mutable objects.
```

Comment:
Don't equate _fewer bugs_ with _bug-free_ :-p

Reply:
`Concurrency` is not an advantage. Imperative languages can have concurrency too.